chore: readme and explanation (#15)
Mean-Finance/subsidies: README.md

> [![Lint](https://github.com/Mean-Finance/subsidies/actions/workflows/lint.yml/badge.svg?branch=main)](https://github.com/Mean-Finance/subsidies/actions/workflows/lint.yml)

Reviewer: Should we rename the repository/package to 'ongoing airdrops' or 'campaigns'?

Author: To be honest, I think we should go for `campaigns`.

[![Tests](https://github.com/Mean-Finance/subsidies/actions/workflows/tests.yml/badge.svg?branch=main)](https://github.com/Mean-Finance/subsidies/actions/workflows/tests.yml)
[![Slither Analysis](https://github.com/Mean-Finance/subsidies/actions/workflows/slither.yml/badge.svg?branch=main)](https://github.com/Mean-Finance/subsidies/actions/workflows/slither.yml)

# Ongoing airdrops / campaigns

> This repository holds the [OngoingAirdrops.sol](./solidity/contracts/OngoingAirdrops.sol) smart contract that is implemented to be able to handle multiple ongoing airdrops, also referred to as **campaigns**.

Reviewer: ```suggestion
This repository holds the [OngoingAirdrops.sol](./solidity/contracts/OngoingAirdrops.sol) smart contract. This contract is meant to handle multiple ongoing airdrops, also referred to as **campaigns**.
```

## Multiple campaigns

> Let's say a protocol wants to do a one-time referral reward in USDC for all users that have done some action. Additionally, this protocol wants to do a one-time airdrop of some loyalty token.

Reviewer: ```suggestion
Let's say a protocol wants to do a one-time referral reward in USDC for all users that have executed some action. Additionally, this protocol wants to do a one-time airdrop of some loyalty token.
```

Instead of having to deploy multiple airdrop claiming smart contracts and making more difficult the task to maintain subgraphs, frontends, etc. they can now create both campaigns on this contract. One might be called `Best swappers` and the other one `Loyalty Program Q1`.

## Ongoing campaigns

> Now the protocol wants to implement an on-going (not a one time) referral campaign that for one reason or another can not be calculated and / or distributed directly / instantly on-chain. Now, the protocol could still want the users to claim this off-chain calculated referral on-chain.

Reviewer: ```suggestion
Another potential use case would be if a protocol wants to implement an on-going (not a one time) referral campaign that for one reason or another cannot be calculated and / or distributed directly on-chain. Instead of having multiple claim contracts, this `OngoingAirdrops` contract can be used instead
```

## Technical details

On-going campaigns work by updating the merkle root whenever we want to add tokens to user allocations.

Merkle root **should** be built by leaves, each of which will be representing a user. This leaf will be calculated by the sum of all previous user allocation plus the new one.

`leaf(U, A) = leaf(U, A - 1) + ... + leaf(A, 1)`

> An exmaple: Let's say Alice had at one time `5 Tokens A` to claim, and now she will have `15 tokens A` to claim, her leaf on the updated root should be `20 tokens A`.

Reviewer: ```suggestion
An example: Let's say Alice was able to claim `5 Tokens A` for the first instance of a campaign, and now she is rewarded `15 tokens A` for the second instance. In this case, her leaf on the updated root should be `20 tokens A`.
```
